rhino-wasm/src/main/java/org/rhino/community/wasm/WebAssembly.java

```diff
@@ -0,0 +1,149 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.rhino.community.wasm;
+
+import com.dylibso.chicory.runtime.Instance;
+import com.dylibso.chicory.runtime.Module;
+import com.dylibso.chicory.wasm.types.Value;
+import org.mozilla.javascript.Callable;
+import org.mozilla.javascript.Context;
+import org.mozilla.javascript.ScriptRuntime;
+import org.mozilla.javascript.Scriptable;
+import org.mozilla.javascript.ScriptableObject;
+import org.mozilla.javascript.Undefined;
+import org.mozilla.javascript.typedarrays.NativeArrayBuffer;
+import org.mozilla.javascript.typedarrays.NativeTypedArrayView;
+
+public class WebAssembly extends ScriptableObject {
+
+    private static final String WEBASSEMBLY_TAG = "WebAssembly";
+
+    public static void init(Context cx, Scriptable scope, boolean sealed) {
+        WebAssembly wasm = new WebAssembly();
+        wasm.setPrototype(getObjectPrototype(scope));
+        wasm.setParentScope(scope);
+
+        wasm.defineProperty(
+                scope, "instantiate", 2, WebAssembly::instantiate, DONTENUM, DONTENUM | READONLY);
+
+        ScriptableObject.defineProperty(scope, WEBASSEMBLY_TAG, wasm, DONTENUM);
+        if (sealed) {
+            wasm.sealObject();
+        }
+    }
+
+    // WebAssembly.instantiate
+    private static Object instantiate(
+            Context cx, Scriptable scope, Scriptable thisObj, Object[] args) {
+        // TODO: implement the other constructors
+        assert (args.length == 1);
+        if (!ScriptRuntime.isObject(thisObj)) {
+            throw ScriptRuntime.typeErrorById("msg.arg.not.object", ScriptRuntime.typeof(thisObj));
+        }
+        Object arg0 = (args.length > 0 ? args[0] : Undefined.instance);
```

Comment:
Inside instantiateInternal it seems arg0 must resolved to a buffer. Shouldn't/couldn't that check be made here?

Reply:
In this case `instantiateInternal` should accept `Object[] args` since those checks are defining the "arity" of the function.
I think I noticed this pattern somewhere in the Rhino codebase, no strong opinions from myself, happy to turn things around as required.